File: practice/mathematics/number_theory/largest_non_coprime_submatrix/largest_non_coprime_submatrix.py
#!/usr/bin/env python

from fractions import gcd

n, m = map(int, raw_input().strip().split())

a = []
for _ in xrange(n):
    row = map(int, raw_input().strip().split())
    a.append(row)

aux = [[[] for _ in xrange(m)] for __ in xrange(n)]
for ROW in xrange(n):
    for COL in xrange(m):
        aux[ROW][COL] = [a[ROW][COL]]
        row_cnt = ROW + 1
        d = a[ROW][COL]
        while row_cnt < n:
            d = gcd(d, a[row_cnt][COL])
            if d == 1:
                break
            aux[ROW][COL].append(d)
            row_cnt += 1

r = 0
for ROW in xrange(n):
    for COL in xrange(m):
        for CNT in xrange(len(aux[ROW][COL]) - 1, -1, -1):
            if (CNT + 1) * (m - COL) < r:
                break
            h = CNT + 1
            col_cnt = COL + 1
            d = aux[ROW][COL][CNT]
            if d > 1:
                w = 1
                while col_cnt < m and CNT < len(aux[ROW][col_cnt]):
                    d = gcd(d, aux[ROW][col_cnt][CNT])
                    if d == 1:
                        break
                    w += 1
                    col_cnt += 1
                if w * h > r:
                    r = w * h
print r

File: practice/mathematics/number_theory/largest_non_coprime_submatrix/largest_non_coprime_submatrix.cpp
#include <vector>
#include <iostream>
#include <algorithm>

using namespace std;

unsigned gcd(unsigned i_a, unsigned i_b)
{
    if (i_a > i_b)
    {
        swap(i_a, i_b);
    }
    
    while (i_a > 0)
    {
        unsigned c = i_b % i_a;
        i_b = i_a;
        i_a = c;
    }
    
    return i_b;
}

int main()
{
    int n, m;
    cin >> n >> m;
    
    vector<vector<unsigned>> a(n, vector<unsigned>(m));
    for (int row = 0; row < n; ++row)
    {
        for (int col = 0; col < m; ++col)
        {
            cin >> a[row][col];
        }
    }
    
    vector<vector<vector<unsigned>>> aux(n, vector<vector<unsigned>>(m));
    for (int row = 0; row < n; ++row)
    {
        for (int col = 0; col < m; ++col)
        {
            aux[row][col].push_back(a[row][col]);
            int rcnt = row + 1;
            unsigned d = a[row][col];
            while (rcnt < n)
            {
                d = gcd(d, a[rcnt][col]);
                if (d == 1)
                {
                    break;
                }
                aux[row][col].push_back(d);
                rcnt++;
            }
        }
    }
    
    unsigned r = 0;
    for (int row = 0; row < n; ++row)
    {
        for (int col = 0; col < m; ++col)
        {
            for (int cnt = aux[row][col].size() - 1; cnt >= 0; --cnt)
            {
                if ((cnt + 1) * (m - col) < r)
                {
                    break;    
                }
                
                int h = cnt + 1;
                int ccnt = col + 1;
                unsigned d = aux[row][col][cnt];
                if (d > 1)
                {
                    unsigned w = 1;
                    while (ccnt < m && cnt < aux[row][ccnt].size())
                    {
                        d = gcd(d, aux[row][ccnt][cnt]);
                        if (d == 1)
                        {
                            break;
                        }
                        w++;
                        ccnt++;
                    }
                    if (w * h > r)
                    {
                        r = w * h;
                    }
                }
            }
        }
    }
    
    cout << r << endl;
    
    return 0;
}

